### PR TITLE
Automated ABI Json tests.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `3323819e55a56b0b3a2f0e920aa04346ab40acb5`:
```
Error: Trailing whitespace found:
 test/libsolidity/ABIJson/basic_test.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/basic_test.sol:16:// "outputs" : 
 test/libsolidity/ABIJson/constructor_abi.sol:7:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:12:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:24:// "outputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:18:// "outputs" : 
 test/libsolidity/ABIJson/event_structs.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/event_structs.sol:14:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:26:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:33:// "components" : 
 test/libsolidity/ABIJson/events.sol:12:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:20:// "outputs" : 
 test/libsolidity/ABIJson/events.sol:33:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:57:// "inputs" : 
 test/libsolidity/ABIJson/function_type.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:34:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:46:// 
 test/libsolidity/ABIJson/inherited.sol:47:// 
 test/libsolidity/ABIJson/inherited.sol:51:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:59:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:72:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:80:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:93:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:106:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:17:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:30:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:38:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:18:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:31:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:39:// "outputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:20:// "outputs" : 
 test/libsolidity/ABIJson/payable_constructor_abi.sol:7:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/pure_function.sol:34:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:42:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:16:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:28:// "inputs" : 
 test/libsolidity/ABIJson/return_structs.sol:14:// "outputs" : 
 test/libsolidity/ABIJson/return_structs.sol:21:// "components" : 
 test/libsolidity/ABIJson/return_structs.sol:28:// "components" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:13:// "outputs" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:16:// "components" : 
 test/libsolidity/ABIJson/structs_and_arrays.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:12:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:15:// "components" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:22:// "components" : 
 test/libsolidity/ABIJson/view_function.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/view_function.sol:34:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:42:// "outputs" : 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `061266d5b66a2e3a0a6f10ab3a2c45913c6a69e3`:
```
Error: Trailing whitespace found:
 test/libsolidity/ABIJson/basic_test.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/basic_test.sol:17:// "outputs" : 
 test/libsolidity/ABIJson/constructor_abi.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:25:// "outputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:19:// "outputs" : 
 test/libsolidity/ABIJson/event_structs.sol:12:// "inputs" : 
 test/libsolidity/ABIJson/event_structs.sol:15:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:27:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:34:// "components" : 
 test/libsolidity/ABIJson/events.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/events.sol:34:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:58:// "inputs" : 
 test/libsolidity/ABIJson/function_type.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:14:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:47:// 
 test/libsolidity/ABIJson/inherited.sol:48:// 
 test/libsolidity/ABIJson/inherited.sol:53:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:61:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:74:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:82:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:95:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:108:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:23:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:18:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:31:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:39:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:19:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:32:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:40:// "outputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/payable_constructor_abi.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/pure_function.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:43:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:17:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:29:// "inputs" : 
 test/libsolidity/ABIJson/return_structs.sol:15:// "outputs" : 
 test/libsolidity/ABIJson/return_structs.sol:22:// "components" : 
 test/libsolidity/ABIJson/return_structs.sol:29:// "components" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:14:// "outputs" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:17:// "components" : 
 test/libsolidity/ABIJson/structs_and_arrays.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:16:// "components" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:23:// "components" : 
 test/libsolidity/ABIJson/view_function.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/view_function.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:43:// "outputs" : 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `07e392ad9f96181d01dd16232e541860681d7853`:
```
Error: Trailing whitespace found:
 test/libsolidity/ABIJson/basic_test.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/basic_test.sol:17:// "outputs" : 
 test/libsolidity/ABIJson/constructor_abi.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_input_parameter_with_named_one.sol:25:// "outputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/empty_name_return_parameters.sol:19:// "outputs" : 
 test/libsolidity/ABIJson/event_structs.sol:12:// "inputs" : 
 test/libsolidity/ABIJson/event_structs.sol:15:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:27:// "components" : 
 test/libsolidity/ABIJson/event_structs.sol:34:// "components" : 
 test/libsolidity/ABIJson/events.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/events.sol:34:// "inputs" : 
 test/libsolidity/ABIJson/events.sol:58:// "inputs" : 
 test/libsolidity/ABIJson/function_type.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:14:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:47:// 
 test/libsolidity/ABIJson/inherited.sol:48:// 
 test/libsolidity/ABIJson/inherited.sol:53:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:61:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:74:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:82:// "outputs" : 
 test/libsolidity/ABIJson/inherited.sol:95:// "inputs" : 
 test/libsolidity/ABIJson/inherited.sol:108:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/library_function.sol:23:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:18:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:31:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods.sol:39:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:11:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:19:// "outputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:32:// "inputs" : 
 test/libsolidity/ABIJson/multiple_methods_order.sol:40:// "outputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:9:// "inputs" : 
 test/libsolidity/ABIJson/multiple_params.sol:21:// "outputs" : 
 test/libsolidity/ABIJson/payable_constructor_abi.sol:8:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/pure_function.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/pure_function.sol:43:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:17:// "outputs" : 
 test/libsolidity/ABIJson/return_param_in_abi.sol:29:// "inputs" : 
 test/libsolidity/ABIJson/return_structs.sol:15:// "outputs" : 
 test/libsolidity/ABIJson/return_structs.sol:22:// "components" : 
 test/libsolidity/ABIJson/return_structs.sol:29:// "components" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:14:// "outputs" : 
 test/libsolidity/ABIJson/return_structs_with_contracts.sol:17:// "components" : 
 test/libsolidity/ABIJson/structs_and_arrays.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:13:// "inputs" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:16:// "components" : 
 test/libsolidity/ABIJson/structs_in_libraries.sol:23:// "components" : 
 test/libsolidity/ABIJson/view_function.sol:10:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:22:// "outputs" : 
 test/libsolidity/ABIJson/view_function.sol:35:// "inputs" : 
 test/libsolidity/ABIJson/view_function.sol:43:// "outputs" : 

```
Please check that your changes are working as intended.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `b4814bd20c42c1dd19bd0e83e57233dcfdc40015`:
```
Error: Trailing whitespace found:
 test/libsolidity/ABIJson/inherited.sol:47:// 
 test/libsolidity/ABIJson/inherited.sol:48:// 

```
Please check that your changes are working as intended.

File: test/libsolidity/ABIJsonTest.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	bool first = true;
+	for (string const& contractName: compiler.contractNames())
+	{
+		if (!first)
+			m_obtainedResult += "\n\n";
+		m_obtainedResult += "    " + contractName + "\n";
+		m_obtainedResult += jsonPrettyPrint(compiler.contractABI(contractName)) + "\n";
+		first = false;
+	}
+	if (m_expectation != m_obtainedResult)
```
Comment (Marenz) [inline test/libsolidity/ABIJsonTest.cpp:69]: Would it make sense to parse the result/expectation and compare the json ast instead?

Comment (Marenz) [reply test/libsolidity/ABIJsonTest.cpp:69]: I guess it would make it easier to highlight the differences in output for example

Comment (chriseth) [reply test/libsolidity/ABIJsonTest.cpp:69]: If we do that, the auto-update might result in a different text even though the comparison resulted in "equal" - is that what we want?

Comment (chriseth) [reply test/libsolidity/ABIJsonTest.cpp:69]: Also I guess it is good to know that this is all deterministic :)

**Review by Marenz — APPROVED** [review]
Once suggestion / comment, but LGTM otherwise.

File: test/libsolidity/ABIJsonTest.cpp (hunk 1 above)
Comment (Marenz) [reply test/libsolidity/ABIJsonTest.cpp:69]: We could do both. And only run the AST comparison once the string comparsion failed ;)

**Pull request merged**